### ISSUE TITLE
cc_shared_library ignores linkopts from header only cc_library deps

### ISSUE THREAD

**Issue by keith** [bazelbuild/bazel#21884]
### Description of the bug:

With this example:

```bzl
cc_library(
    name = "foo",
    srcs = ["foo.cpp"],
    linkopts = ["-lcompression"],
)

cc_library(
    name = "baz",
    hdrs = ["baz.hpp"],
    linkopts = ["-blah"],
)

cc_binary(
    name = "main",
    srcs = ["main.cpp"],
    deps = [
        ":baz",
        ":foo",
    ],
)

cc_shared_library(
    name = "bar",
    deps = [
        ":baz",
        ":foo",
    ],
)
```

The cc_binary target fails to build, but the cc_shared_library doesn't, it ignores the invalid `-blah` linkopt. If you add an empty cpp file to `baz` it works correctly.

### Which operating system are you running Bazel on?

macOS

### What is the output of `bazel info release`?

7.1.1

**Comment by comius** [comment]
cc @pzembrod for triage
cc @oquenchil for potential insights

**Comment by keith** [comment]
https://github.com/bazelbuild/bazel/issues/23053

**Comment by iancha1992** [comment]
@bazel-io fork 7.4.0

**Comment by keith** [comment]
fix PR is here https://github.com/bazelbuild/bazel/pull/23215

**Comment by iancha1992** [comment]
@bazel-io fork 8.0.0

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.4.0 RC3](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc3). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc3`. Thanks!

**Comment by Xiangze-Li** [comment]
I see the fix #24017 (in 7.4.0) had been reverted in #24087 (in 7.4.1). Is this issue still planned to be fixed in Bazel 7?

I am using a precompiled static library, and that library on Windows needs to be linked with some Windows library. So I write a header only `cc_library` with `linkopts` like:

```starlark
cc_import(
  name = "foo_static",
  static_library = "precompiled.lib"
)

cc_library(
  name = "foo",
  hdrs = ["foo.h"],
  linkopts = ["-DEFAULTLIB:gdi32.lib"],
  deps = [":foo_static"]
)
```

Then the `foo` is used by some `cc_shared_library`, with `linkopts` dropped due to this issue.

**Comment by fmeum** [comment]
The fix uncovered other bugs that are much harder to fix. I'm still planning to roll it forward, but fixing the other bug will take some time and thought.
